### PR TITLE
Use own psalm instead of a global one

### PR DESCRIPTION
**Pull request by ChristophWurst**
Our CI used a global psalm and not one we install via composer. Hence this is waiting for random CI failure when the one in the docker image updates. We should have control over the version and use Dependabot for updates to make the manageable.

- [x] Requires https://github.com/nextcloud/server/pull/23409
- [x] Requires https://github.com/nextcloud/server/pull/23410

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -17,11 +17,12 @@
     },
     "require-dev": {
         "nextcloud/coding-standard": "^0.3.0",
-        "psalm/phar": "^3.12"
+        "vimeo/psalm": "^3.17"
```
Comment (kesselb) [inline composer.json:20]: An advantage of the phar version is less troubles with dependencies.

**Review by kesselb — APPROVED** [review]

File: composer.json (hunk 1 above)
Comment (ChristophWurst) [reply composer.json:20]: It is indeed. But so far there are no conflicts so I'd use the "real" one as long as possible :wink:

Comment (kesselb) [reply composer.json]: Good ;) We install a lot of dependencies (like symfony/console) which are already part of 3rdparty. But lib/composer is ignored so this should work.

I would suggest to migrate the composer dev-dependencies to phive (https://phar.io/) at some point. That's more convenient to setup.

**Comment by MorrisJobke** [comment]
> * [x]  Requires #23409
> * [x]  Requires #23410

Both are merged ✅

File: .github/workflows/static-code-analysis.yml [hunk 2]
```diff
@@ -13,13 +13,17 @@ jobs:
...
               with:
-                args: psalm --monochrome --no-progress --output-format=text --update-baseline || ( git diff && exit 1 )
+                php-version: 7.4
+                coverage: none
+            - name: Composer install
+              run: composer i
+            - name: Psalm
+              run: composer run psalm -- --monochrome --no-progress --output-format=text --update-baseline || ( git diff -- . ':!lib/composer' && exit 1 )
+            - name: Check diff
+              run: git diff -- . ':!lib/composer' && exit 1
```
Comment (MorrisJobke) [inline .github/workflows/static-code-analysis.yml:26]:
```suggestion
              run: git diff -- . ':!lib/composer'
```

We should not error on fixed issues. There is a daily GitHub workflow that commits this.

File: .github/workflows/static-code-analysis.yml [hunk 3]
```diff
@@ -13,13 +13,17 @@ jobs:
...
               with:
-                args: psalm --monochrome --no-progress --output-format=text --update-baseline || ( git diff && exit 1 )
+                php-version: 7.4
+                coverage: none
+            - name: Composer install
+              run: composer i
+            - name: Psalm
+              run: composer run psalm -- --monochrome --no-progress --output-format=text --update-baseline || ( git diff -- . ':!lib/composer' && exit 1 )
+            - name: Check diff
+              run: git diff -- . ':!lib/composer'
```
Comment (MorrisJobke) [inline .github/workflows/static-code-analysis.yml:26]:
```suggestion
```

**Review by MorrisJobke — APPROVED** [review]

**Comment by MorrisJobke** [comment]
Backport to stable20? It started failing in https://github.com/nextcloud/server/pull/23466/checks?check_run_id=1258325103 😢

**Comment by MorrisJobke** [comment]
I'm also fine with adding more stuff to the baseline in stable20, but it should not fail.

**Pull request merged**